docs/development/abi.md

```diff
@@ -0,0 +1,221 @@
+- `libwebsocket.js`
+- `bzip2`
+- `zlib`
+
+All of these come from Emscripten ports and the versions of these libraries are
+determined by the version of Emscripten we build with. Any symbols from these
+static libraries may be used by shared libraries.
+
+### pyodide_2024_0
+
+This ABI is intended for use with Python 3.12. The Emscripten version is 3.1.58.
+
+All shared libraries must be linked with `-sWASM_BIGINT`.
+
+By default, C++ libraries are built with exceptions disabled, and `throw` is an
+abort. The same is true for `setjmp`/`longjmp`. To enable exceptions and
+`setjmp`/`longjmp`, `-fexceptions` must be passed at compile time and link time.
+
+Specifying an RPATH is not supported and will cause link errors. The dynamic
+loader has been patched so that all dynamic libraries in a wheel named
+`wheel_name-<tag>.whl` will be loaded as if
+`/lib/python3.12/site-packages/wheel_name.libs` is on the `RPATH`, so any
+dynamic library dependencies should be placed in the wheel in a folder called
+`wheel_name.libs`.
+
+### pyodide_2025_0 (under development)
+
+This section reflects the aspirational ABI for `pyodide_2025_0`. This is all
+subject to change without notice.
+
+This ABI is intended for use with Python 3.13. The Emscripten version is 4.0.6.
+
+By default, emcc will compile and link C++ libraries with exceptions disabled,
+so that `throw` is an abort. The same is true for `setjmp`/`longjmp`. To enable
+exceptions and `setjmp`/`longjmp`, `-fwasm-exceptions` must be passed to `emcc`
+at compile time and link time.
+
+There is full support for `RPATH`. The dynamic loader will only load
+dependencies that are properly specified on the `RPATH`, just being in
+`/lib/python3.13/site-packages/wheel_name.libs` is not sufficient.
+
+For building Rust, it is necessary to use a Rust nightly after January 15th, 2025.
+The flag `-Z emscripten-wasm-eh` must be passed. It is also necessary to use a
+compatible Emscripten sysroot that has been built with WebAssembly exception handling.
+Such sysroots are produced and distributed by
+[pyodide/rust-emscripten-wasm-eh-sysroot](https://github.com/pyodide/rust-emscripten-wasm-eh-sysroot).
+This is only distributed for Rust nightly-2025-02-01. To use a different Rust
+nighly, it is possible to clone the `pyodide/rust-emscripten-wasm-eh-sysroot`
+repository and follow the instructions in the README to build a compatible
+sysroot. To install the Emscripten sysroot use:
+
+```sh
+rustup toolchain install nightly-2025-02-01
+TOOLCHAIN_ROOT=$(rustup which --toolchain nightly-2025-02-01 rustc)
+RUSTLIB=$TOOLCHAIN_ROOT/lib/rustlib
+wget https://github.com/pyodide/rust-emscripten-wasm-eh-sysroot/releases/download/emcc-4.0.6_nightly-2025-02-01/emcc-4.0.6_nightly-2025-02-01.tar.bz2
+tar -xf emcc-4.0.6_nightly-2025-02-01.tar.bz2 --directory=$RUSTLIB
+```
+
+Note that this is all necessary _even if_ the crate uses `-Cpanic=abort` because
+the Rust standard library is not built with `-Cpanic=abort` and so there will be
```

Comment:
You can rebuild std with `-Z build-std=std,panic_abort -Z build-std-features=panic_immediate_abort` to enable panic=abort inside std as well

Reply:
I've had trouble with making `-Z build-std -C panic=abort` work at all.

Reply:
Maybe `-Z build-std-features=panic_immediate_abort` is what you need to fix that. But my feeling `-Z build-std` is too buggy to be usable.

Reply:
But I will mention the possibility of using `-Zbuild-std` instead.